packager/packager.cc

```diff
@@ -479,34 +517,200 @@ ast::UnresolvedConstantLit *verifyConstant(core::MutableContext ctx, core::NameR
-    return std::equal(a.begin(), a.begin() + minSize, b.begin(), b.begin() + minSize);
-}
+// Interface for traversing the tree of package namespaces.
+// Compactly represent current position in this tree with begin/end offsets. This relies on
+// lexicographic sorting of the packages vector. Push operations take advantage of binary search
+// to be O(log(end - begin)).
+//
+// For example with names=[Foo, Bar]
+// 0 Foo
+// 1 Foo::Bar::Baz    <-- begin
+// 2 Foo::Bar::Blub
+// 3 Foo::Buzz        <-- end
+// 4 Quuz::Bang
+// 5 Yaz
+// 6 <end of list>
+class PackageNamespaces final {
+    using Bound = pair<uint16_t, uint16_t>;
+
+    const vector<core::NameRef> &packages; // Mangled names sorted lexicographically
+    const PackageInfoImpl &filePkg;        // Package for current file
+    // Current bounds:
+    uint16_t begin;
+    uint16_t end;
+
+    const bool isTestFile;
+    const uint16_t filePkgIdx;
+
+    int skips = 0;
+    vector<Bound> bounds;
+    vector<core::NameRef> nameParts;
+    vector<pair<core::NameRef, uint16_t>> curPkg;
+    bool foundTestNS = false;
+
+    static constexpr uint16_t SKIP_BOUND_VAL = 0;
+
+public:
+    PackageNamespaces(core::Context ctx, const PackageInfoImpl &filePkg, bool isTestFile)
+        : packages(ctx.state.packageDB().packages()), filePkg(filePkg), begin(0), end(packages.size()),
+          isTestFile(isTestFile), filePkgIdx(findPackageIndex(ctx, filePkg)) {
+        ENFORCE(packages.size() < numeric_limits<uint16_t>::max());
+    }
+
+    int depth() const {
+        return nameParts.size();
+    }
+
+    const vector<core::NameRef> &currentConstantName() const {
+        return nameParts;
+    }
+
+    core::NameRef packageForNamespace(core::Context ctx) const {
+        if (curPkg.empty()) {
+            return core::NameRef::noName();
+        }
+        return curPkg.back().first;
+    }
+
+    bool onPackagePath(core::Context ctx) {
+        if (begin <= filePkgIdx && filePkgIdx < end) {
+            return true;
+        }
+        if (!curPkg.empty() && curPkg.back().first == filePkg.mangledName()) {
+            return true;
+        }
+        return false;
+    }
+
+    void pushName(core::Context ctx, core::NameRef name) {
+        if (skips > 0) {
+            skips++;
+            return;
+        }
+        uint16_t idx = bounds.size();
+
+        if (isTestFile && idx == 0 && !foundTestNS) {
+            if (isPrimaryTestNamespace(name)) {
+                foundTestNS = true;
+                return;
+            } else if (!isTestNamespace(ctx, name)) {
+                // Inside a test file, but not inside a test namespace. Set bounds such that
+                // begin == end, stopping any subsequent search.
+                bounds.emplace_back(begin, end);
+                nameParts.emplace_back(name);
+                begin = end = 0;
+                return;
+            }
+        }
+
+        if (idx > 0 && end - begin == 1 && packages[begin] == filePkg.mangledName()) {
```

Comment:
`idx` was only used to check if bounds was empty.
```suggestion
        uint16_t boundsEmpty = bounds.empty();

        if (isTestFile && boundsEmpty && !foundTestNS) {
            if (isPrimaryTestNamespace(name)) {
                foundTestNS = true;
                return;
            } else if (!isTestNamespace(ctx, name)) {
                // Inside a test file, but not inside a test namespace. Set bounds such that
                // begin == end, stopping any subsequent search.
                bounds.emplace_back(begin, end);
                nameParts.emplace_back(name);
                begin = end = 0;
                return;
            }
        }

        if (!boundsEmpty && end - begin == 1 && packages[begin] == filePkg.mangledName()) {
```